### PR TITLE
IDEMPIERE-2981 - Implement JSON field type

### PR DESCRIPTION
**Pull request by d-ruiz**
https://idempiere.atlassian.net/browse/IDEMPIERE-2981

# Pull Request Checklist

- [x] My code follows the [code guidelines](https://wiki.idempiere.org/en/Contributing_to_Trunk) of this project
- [x] My code follows the best practices of this project
- [x] I have performed a self-review of my own code
- [x] My code is easy to understand and review. 
- [x] I have checked my code and corrected any misspellings
- [x] In hard-to-understand areas, I have commented my code.
- [x] My changes generate no new warnings
### Tests
- [x] I have tested the direct scenario that my code is solving
- [x] I checked all collaterals that can be affected by my changes, and tested other potential affected scenarios
- [x] New and existing unit tests pass locally with my changes
- [x] I have added unit tests that prove my fix is effective or that my feature works
### Documentation
- [ ] I have made corresponding changes to the documentation as follows:
- - [ ] New feature (non-breaking change which adds functionality): I have created the New Feature page in the project wiki explaining the functionality and how to use it. If relevant, I have committed sample data to the core seed to have usable examples in GardenWorld.
- - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected): I have documented the change in a clear way that everyone in the community can understand the impact of the change.
- - [ ] Improvement (improves and existing functionality): This documentation is needed if the improvement changes the way the user interacts with the system or the outcome of a process/task changes. If it is just, for instance, a performance improvement, documentation might not be needed. 
- [ ] The changed/added documentation is in the project wiki (not privately-hosted pdf files or links pointing to a company website) and is complete and self-explanatory.

### CODE REVIEW

File: org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java [hunk 1]
```diff
@@ -536,7 +536,30 @@ public String TO_NUMBER (BigDecimal number, int displayType)
         }
         return result.toString();
     }   //  TO_NUMBER
+    
+	/**
+	 *	@return string with right casting for JSON inserts
+	 */
+	public String getJSONCast () {
+		return "parse_json(?)";
```
Comment (hengsin) [inline org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java:544]: not 100% sure but it seems to me parse_json is for Oracle NoSQL DB ...

File: org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java [hunk 2]
```diff
@@ -1026,6 +1049,11 @@ public String getBlobDataType() {
 	public String getClobDataType() {
 		return "CLOB";
 	}
+	
+	@Override
+	public String getJsonDataType() {
+		return "JSON";
```
Comment (hengsin) [inline org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java:1055]: this require Oracle 21c and above.  before 21c, it is usually store as varchar or clob and use the check constraint (columnName is JSON).

File: migration/iD12/oracle/202402261354_IDEMPIERE-2981.sql [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+;
+
+-- Feb 26, 2024, 1:55:37 PM CET
+INSERT INTO AD_Column (AD_Column_ID,Version,Name,Description,AD_Table_ID,ColumnName,FieldLength,IsKey,IsParent,IsMandatory,IsTranslated,IsIdentifier,SeqNo,IsEncrypted,AD_Reference_ID,AD_Client_ID,AD_Org_ID,IsActive,Created,CreatedBy,Updated,UpdatedBy,AD_Element_ID,IsUpdateable,IsSelectionColumn,EntityType,IsSyncDatabase,IsAlwaysUpdateable,IsAutocomplete,IsAllowLogging,AD_Column_UU,IsAllowCopy,SeqNoSelection,IsToolbarButton,IsSecure,IsHtml,IsPartitionKey) VALUES (216570,0,'JSON Data','The json field stores json data.',135,'JsonData',2000,'N','N','N','N','N',0,'N',200267,0,0,'Y',TO_TIMESTAMP('2024-02-26 13:55:36','YYYY-MM-DD HH24:MI:SS'),100,TO_TIMESTAMP('2024-02-26 13:55:36','YYYY-MM-DD HH24:MI:SS'),100,203924,'Y','N','D','N','N','N','Y','927b83df-d161-4332-ad44-8ffed99e8cf4','Y',0,'N','N','N','N')
```
Comment (CarlosRuiz-globalqss) [inline migration/iD12/oracle/202402261354_IDEMPIERE-2981.sql:12]: The scripts don't have the "ALTER TABLE Test" that must correspond to this INSERT

Comment (d-ruiz) [reply migration/iD12/oracle/202402261354_IDEMPIERE-2981.sql:12]: Fixed it with a new commit

File: org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java [hunk 4]
```diff
@@ -536,7 +536,30 @@ public String TO_NUMBER (BigDecimal number, int displayType)
...
+	 *	@param value
+	 *	@return value as json
+	 */
+	public String TO_JSON (String value)
+	{
+		if (value == null)
+			return "NULL";
+
+		StringBuilder retValue = null;
+		retValue = new StringBuilder("parse_json(");
```
Comment (hengsin) [inline org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java:558]: this should be "json(value)", not parse_json(value)

File: org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java (hunk 1 above)
Comment (hengsin) [reply org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java:544]: should be "json(?)" instead.

Comment (d-ruiz) [reply org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java]: changed in a new commit

**Review by CarlosRuiz-globalqss — CHANGES_REQUESTED** [review]
@d-ruiz 

Can you please apply the following patch:
[IDEMPIERE-2981_PR.patch.txt](https://github.com/idempiere/idempiere/files/14524404/IDEMPIERE-2981_PR.patch.txt)

With this patch I tested in postgresql and oracle the following:
* save a JSON value
* change a JSON value
* create new col
* changelog (not in oracle because is CLOB)
* packout
* packin
* report
* export
* CSV import

the unit test is also successful in oracle and postgresql after the changes

**Comment by d-ruiz** [comment]
Patch applied

**Review by CarlosRuiz-globalqss — APPROVED** [review]
Hi @hengsin 


I conducted tests with postgresql and oracle 12 successfully

do you want to check something additional?

**Comment by hengsin** [comment]
No, you can go ahead to merge this.

**Pull request merged**